### PR TITLE
raftstore: remove stale peer as soon as possible

### PR DESCRIPTION
**Pull request by BusyJay**
When a peer is applying snapshot, it won't have any tasks in apply worker, hence it's safe to destroy it immediately.

We need fail point support to write a stable tests.

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -1160,10 +1168,19 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-        info!("[region {}] destroy peer {:?}", region_id, peer);
-        // TODO: should we check None here?
         // Can we destroy it in another thread later?
-        let mut p = self.region_peers.remove(&region_id).unwrap();
+        let mut p = match self.region_peers.remove(&region_id) {
+            None => return,
+            Some(p) => if p.peer_id() == peer.get_id() {
+                p
+            } else {
+                // It has been destroyed.
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: need to check p.peer_id() must > peer.get_id() here?

**Comment by javaforfun** [comment]
LGTM

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -1160,10 +1168,20 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
-        info!("[region {}] destroy peer {:?}", region_id, peer);
-        // TODO: should we check None here?
         // Can we destroy it in another thread later?
-        let mut p = self.region_peers.remove(&region_id).unwrap();
+        let mut p = match self.region_peers.remove(&region_id) {
+            None => return,
+            Some(p) => if p.peer_id() == peer.get_id() {
+                p
+            } else {
+                assert!(p.peer_id() > peer.get_id());
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: add a comment for how can we meet this case.

**Pull request merged**